File: benchmarks/python/perf_compare.py
from __future__ import annotations

import copy
import json
import pathlib
import sys
import time
from typing import Callable


ROOT = pathlib.Path(__file__).resolve().parents[2]
PYTHON_CAN = ROOT / ".external" / "python-can"
sys.path.insert(0, str(PYTHON_CAN))

import can  # noqa: E402


def bench(name: str, iterations: int, func: Callable[[], object]) -> dict[str, object]:
    start = time.perf_counter_ns()
    for _ in range(iterations):
        func()
    total_ns = time.perf_counter_ns() - start
    return {
        "name": name,
        "iterations": iterations,
        "total_ns": total_ns,
        "ns_per_iter": total_ns / iterations,
    }


def main() -> None:
    iterations = int(sys.argv[1]) if len(sys.argv) > 1 else 1_000_000
    classic_payload = bytes([1, 2, 3, 4, 5, 6, 7, 8])
    fd_payload = bytes([0xAA] * 64)
    clone_msg = can.Message(
        arbitration_id=0x123,
        is_extended_id=False,
        data=classic_payload,
    )
    validate_msg = can.Message(
        arbitration_id=0x123,
        is_extended_id=False,
        data=classic_payload,
    )
    filter_msg = can.Message(
        arbitration_id=0x18FF_50E5,
        is_extended_id=True,
        data=classic_payload,
    )
    bus = can.Bus(
        interface="virtual",
        channel=f"rust-can-bench-{id(filter_msg)}",
        receive_own_messages=True,
    )
    bus.set_filters(
        [
            {"can_id": 0x100, "can_mask": 0x700, "extended": False},
            {"can_id": 0x200, "can_mask": 0x700, "extended": False},
            {"can_id": 0x18FF_0000, "can_mask": 0x1FFF_0000, "extended": True},
            {"can_id": 0x7E8, "can_mask": 0x7FF, "extended": False},
        ]
    )

    try:
        results = [
            bench(
                "classic_message_create_8b",
                iterations,
                lambda: can.Message(
                    arbitration_id=0x123,
                    is_extended_id=False,
                    data=classic_payload,
                ),
            ),
            bench(
                "fd_message_create_64b",
                iterations,
                lambda: can.Message(
                    arbitration_id=0x18FF_50E5,
                    is_extended_id=True,
                    is_fd=True,
                    bitrate_switch=True,
                    data=fd_payload,
                ),
            ),
            bench("message_clone_8b", iterations, lambda: copy.copy(clone_msg)),
            bench("message_validate_8b", iterations, lambda: validate_msg._check()),
            bench("filter_match_4_filters", iterations, lambda: bus._matches_filters(filter_msg)),
        ]
    finally:
        bus.shutdown()

    print(
        json.dumps(
            {
                "language": "python",
                "python_can_commit": "491a691fd1faffab1c48956bafd711e7c653db54",
                "iterations": iterations,
                "results": results,
            },
            indent=2,
        )
    )


if __name__ == "__main__":
    main()

File: benchmarks/src/bin/perf_compare.rs
//! Emit machine-readable microbenchmark results for core message operations.

use std::hint::black_box;
use std::time::Instant;

use rust_can_core::filter::{CanFilter, CanFilters};
use rust_can_core::message::CanMessage;

struct BenchResult {
    name: &'static str,
    iterations: u64,
    total_ns: u128,
}

impl BenchResult {
    fn ns_per_iter(&self) -> f64 {
        self.total_ns as f64 / self.iterations as f64
    }
}

fn bench<F>(name: &'static str, iterations: u64, mut f: F) -> BenchResult
where
    F: FnMut(),
{
    let start = Instant::now();
    for _ in 0..iterations {
        f();
    }
    BenchResult {
        name,
        iterations,
        total_ns: start.elapsed().as_nanos(),
    }
}

fn main() {
    let iterations = std::env::args()
        .nth(1)
        .and_then(|arg| arg.parse::<u64>().ok())
        .unwrap_or(1_000_000);

    let classic_payload = [1_u8, 2, 3, 4, 5, 6, 7, 8];
    let fd_payload = [0xAA_u8; 64];
    let clone_msg = CanMessage::new(0x123, &classic_payload, false).unwrap();
    let validate_msg = CanMessage::new(0x123, &classic_payload, false).unwrap();
    let filter_msg = CanMessage::new(0x18FF_50E5, &classic_payload, true).unwrap();
    let filters = CanFilters::from_filters(vec![
        CanFilter::new(0x100, 0x700, Some(false)),
        CanFilter::new(0x200, 0x700, Some(false)),
        CanFilter::new(0x18FF_0000, 0x1FFF_0000, Some(true)),
        CanFilter::new(0x7E8, 0x7FF, Some(false)),
    ]);

    let results = [
        bench("classic_message_create_8b", iterations, || {
            black_box(CanMessage::new(0x123, black_box(&classic_payload), false).unwrap());
        }),
        bench("fd_message_create_64b", iterations, || {
            black_box(CanMessage::new_fd(
                0x18FF_50E5,
                black_box(&fd_payload),
                true,
                true,
            )
            .unwrap());
        }),
        bench("message_clone_8b", iterations, || {
            black_box(black_box(&clone_msg).clone());
        }),
        bench("message_validate_8b", iterations, || {
            black_box(&validate_msg).validate().unwrap();
            black_box(());
        }),
        bench("filter_match_4_filters", iterations, || {
            black_box(black_box(&filters).matches(black_box(&filter_msg)));
        }),
    ];

    println!("{{");
    println!("  \"language\": \"rust\",");
    println!("  \"iterations\": {},", iterations);
    println!("  \"results\": [");
    for (idx, result) in results.iter().enumerate() {
        let comma = if idx + 1 == results.len() { "" } else { "," };
        println!(
            "    {{\"name\":\"{}\",\"iterations\":{},\"total_ns\":{},\"ns_per_iter\":{:.3}}}{}",
            result.name,
            result.iterations,
            result.total_ns,
            result.ns_per_iter(),
            comma
        );
    }
    println!("  ]");
    println!("}}");
}

#[cfg(test)]
mod tests {
    use super::{bench, BenchResult};

    #[test]
    fn bench_result_reports_nanoseconds_per_iteration() {
        let result = BenchResult {
            name: "unit",
            iterations: 4,
            total_ns: 20,
        };

        assert_eq!(result.ns_per_iter(), 5.0);
        assert_eq!(result.name, "unit");
    }

    #[test]
    fn bench_invokes_closure_for_each_iteration() {
        let mut calls = 0_u64;
        let result = bench("count", 3, || {
            calls += 1;
        });

        assert_eq!(calls, 3);
        assert_eq!(result.name, "count");
        assert_eq!(result.iterations, 3);
    }
}
